spec/controllers/grade_entry_forms_controller_spec.rb

```diff
@@ -0,0 +1,124 @@
+                    upload: { grades_file: @file_without_extension }
+      expect(response.status).to eq(302)
+      expect(flash[:error]).to_not be_empty
+      expect(response).to redirect_to(
+        grades_grade_entry_form_path(grade_entry_form, locale: 'en'))
+    end
+
+    it 'does not accept fileless submission' do
+      post :csv_upload, id: grade_entry_form
+      expect(response.status).to eq(302)
+      expect(flash[:error]).to_not be_empty
+      expect(response).to redirect_to(
+        grades_grade_entry_form_path(grade_entry_form, locale: 'en'))
+    end
+    
+    # this test is currently failing
+    # issue #2075 has been opened to resolve this
+    # it 'should gracefully fail on non-csv file with .csv extension' do
+    #  post :csv_upload, id: grade_entry_form,
+    #      upload: { grades_file: @file_bad_csv }
+    # expect(response.status).to eq(302)
+    # expect(flash[:error]).to_not be_empty
+    # expect(response).to redirect_to(
+    #   grades_grade_entry_form_path(grade_entry_form, locale: 'en'))
+    # end
+    
+    it 'should gracefully fail on .xls file' do
+      post :csv_upload, id: grade_entry_form,
+                    upload: { grades_file: @file_wrong_format }
+      expect(response.status).to eq(302)
+      expect(flash[:error]).to_not be_empty
+      expect(response).to redirect_to(
+        grades_grade_entry_form_path(grade_entry_form, locale: 'en'))
+    end
+  end
+
+  context 'CSV_Downloads' do
+    let(:csv_data) { grade_entry_form.get_csv_grades_report }
+    let(:csv_options) { {
+      filename: "#{grade_entry_form.short_identifier}_grades_report.csv",
+      disposition: 'attachment',
+      type: 'application/vnd.ms-excel'
+    } }
+
+    it 'tests that action csv_downloads returns OK' do
+      get :csv_download, id: grade_entry_form
+      expect(response.status).to eq(200)
+    end
+
+    it 'expects a call to send_data' do
+      expect(@controller).to receive(:send_data).with(csv_data, csv_options) {
+        @controller.render nothing: true #to prevent a 'missing template' error
+      }
+      get :csv_download, id: grade_entry_form
+    end
+
+    #parse header object to check for the right disposition
+    it 'sets disposition as attachment' do
+      get :csv_download, id: grade_entry_form
+      d = response.header['Content-Disposition'].split.first
+      expect(d).to eq 'attachment;'
+    end
+
+    #parse header object to check for the right content type
+    it 'returns vnd.ms-excel type' do
+      get :csv_download, id: grade_entry_form
+      expect(response.content_type).to eq 'application/vnd.ms-excel'
+    end
+
+    #parse header object to check for the right file naming convention
+    it 'filename passes naming conventions' do
+      get :csv_download, id: grade_entry_form
+      filename = response.header['Content-Disposition']
+        .split.last.split('"').second
+      expect(filename).to eq "#{grade_entry_form.short_identifier}" +
+        "_grades_report.csv"
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.